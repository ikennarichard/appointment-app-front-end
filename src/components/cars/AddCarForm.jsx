import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { addCar } from '../../redux/cars/apiSlice';

export default function AddCarForm() {
  const resourceOwner = useSelector((state) => state.auth.resource_owner);
  const { message, error } = useSelector((state) => state.cars);

  const [carDetails, setCarDetails] = useState({
    car_model: '',
    description: '',
    photo: '',
    reservation_price: 0,
    user_id: resourceOwner.id,
  });

  const navigate = useNavigate();

  const dispatch = useDispatch();
  const userId = resourceOwner.id;

  const handleChange = (e) => {
    const { name, value } = e.target;
    setCarDetails((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    dispatch(addCar({ userId, details: carDetails }));
  };

  useEffect(() => {
    if (message === 'Car was added successfully') navigate('/');
  }, [message, navigate, error]);

  return (
    <div>
      <h2>Create a New Car</h2>
      <form onSubmit={handleSubmit}>
        <div className="field">
          <label htmlFor="car_model">
            Car Model:
            <input
              type="text"
              name="car_model"
              value={carDetails.car_model}
              onChange={handleChange}
              id="car_model"
              required
            />
          </label>
        </div>
        <br />
        <div className="field">
          <label htmlFor="description">
            Description:
            <textarea
              name="description"
              value={carDetails.description}
              onChange={handleChange}
              id="description"
              required
            />
          </label>
        </div>
        <br />
        <div className="field">
          <label htmlFor="photo">
            Photo URL:
            <input
              type="text"
              name="photo"
              value={carDetails.photo}
              onChange={handleChange}
              placeholder="enter a valide image url"
              id="photo"
              required
            />
          </label>
        </div>
        <div className="field">
          <label htmlFor="reservation_price">
            Reservation Price:
            <input
              type="number"
              name="reservation_price"
              value={carDetails.reservation_price}
              onChange={handleChange}
              required
            />
          </label>
        </div>
        <button type="submit">Add Car</button>
      </form>
    </div>
  );
}
