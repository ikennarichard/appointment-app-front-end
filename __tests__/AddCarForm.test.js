import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { useDispatch as useDispatchMock, useSelector as useSelectorMock } from 'react-redux';
import { useNavigate as useNavigateMock, MemoryRouter } from 'react-router-dom';
import AddCarForm from '../src/components/cars/AddCarForm';

jest.mock('react-redux');
jest.mock('react-router-dom');

const useDispatch = useDispatchMock;
const useSelector = useSelectorMock;
const navigate = useNavigateMock;

describe('AddCarForm component', () => {
  const renderFormComponent = (mockState) => {
    useSelector.mockReturnValue(mockState);
    useDispatch.mockReturnValue(jest.fn());
    navigate.mockReturnValue(jest.fn());
    render(<AddCarForm />, { wrapper: MemoryRouter });
  };

  it('matches snapshot', () => {
    renderFormComponent({
      cars: [],
      loading: false,
      message: '',
      error: null,
      auth: { resource_owner: 1 },
    });

    const tree = renderer.create(<AddCarForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
