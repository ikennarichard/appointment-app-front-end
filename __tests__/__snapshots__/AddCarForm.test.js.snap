// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCarForm component matches snapshot 1`] = `
<div
  className="d-flex flex-column align-items-center mt-5"
  id="new-car-container"
>
  <div
    className="messages"
  />
  <h1
    className="text-center display-5 fw-bold m-0"
  >
    Create a New Car
  </h1>
  <form
    className="form-container w-100 d-flex justify-content-center mt-2"
    onSubmit={[Function]}
  >
    <div
      className="d-flex flex-column gap-3 w-75 form-inputs-container"
    >
      <div
        className="field"
      >
        <input
          className="p-2 w-100 rounded-1 border-0"
          id="car_model"
          name="car_model"
          onChange={[Function]}
          placeholder="Enter Car Model"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <textarea
          className="p-2 w-100 rounded-1 border-0"
          id="description"
          name="description"
          onChange={[Function]}
          placeholder="Enter Description"
          required={true}
          value=""
        />
      </div>
      <div
        className="field"
      >
        <input
          className="p-2 w-100 rounded-1 border-0"
          id="photo"
          name="photo"
          onChange={[Function]}
          placeholder="Enter a Valid Image URL"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <input
          className="p-2 w-100 rounded-1 border-0"
          id="reservation_price"
          name="reservation_price"
          onChange={[Function]}
          placeholder="Enter Reservation Price"
          required={true}
          type="number"
          value={0}
        />
      </div>
      <button
        className="p-2 w-100 rounded-1 border-0 add-car-btn"
        type="submit"
      >
        Add Car
      </button>
    </div>
  </form>
</div>
`;
